lib/fluent/plugin/base.rb

```diff
@@ -52,14 +52,12 @@ def fluentd_worker_id
         @_fluentd_worker_id
       end
 
+      # @param conf [Fluent::Config::Element, Hash] configuration
       def configure(conf)
-        if Fluent::Engine.supervisor_mode || (conf.respond_to?(:for_this_worker?) && conf.for_this_worker?)
-          workers = if conf.target_worker_ids && !conf.target_worker_ids.empty?
-                      conf.target_worker_ids.size
-                    else
-                      1
-                    end
-          system_config_override(workers: workers)
+        if conf.is_a?(Fluent::Config::Element)
```

Comment:
> `conf` of `Fluent::TextParser.configure(conf)` is `Hash`.
> Thus, I added a `conf.is_a?(Fluent::Config::Element)` check.

Thanks for the information.
This made me notice that API compatibility features for v0.12 may be involved.

* https://docs.fluentd.org/plugin-development#overview
* https://www.shigemk2.com/entry/memo_Fluentd_v0.14_Plugin_API_Updates

I have looked into this, but I am not sure whether we should really assume that the type of `conf` is not `Fluent::Config::Element`.

Some test codes look like passing `Hash` to `Fluent::Plugin::Base::configure()`.
(We need to be careful not to confuse it with `Fluent::Test::Driver::Base::configure()`)
Perhaps these tests are related to the v0.12 compatible layer.

* https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/test/test_formatter.rb
  * https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/test/test_formatter.rb#L317-L321
  * https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/test/test_formatter.rb#L332-L334
  * https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/test/test_formatter.rb#L339-L341
* https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/test/compat/test_parser.rb
  * Entirely.

However, it appears that `Hash` is not passed unless we explicitly pass it in tests.
(If `Hash` can be passed during normal operation, please let me know.)

https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/supervisor.rb#L808-L814
https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/config.rb#L68-L79
https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/config/v1_parser.rb#L43-L54
https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/config/parser.rb#L25-L28
https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/config/dsl.rb#L58-L62
https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/config/element.rb#L67-L72
https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/config/element.rb#L93-L97
https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/root_agent.rb#L155-L163
https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/plugin_helper/parser.rb#L40-L52
https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/plugin_helper/compat_parameters.rb#L304-L305

If so, I thought we could modify the tests and limit the type to `Element`.
But we should consider whether it is acceptable to break backward compatibility, even though this is a test-only impact.
Or it might be possible to fix it without breaking backward compatibility at all. (I haven't checked it yet.)

In the first place, I'm also beginning to wonder whether it's correct to overwrite `workers` in `Fluent::Plugin::Base::configure()`.
There may be a more appropriate place and timing, such as `Supervisor` or `RootAgent`.

I'll think about it a little more.

I would like to figure out what we should do about these problems in the future and then figure out how far we should go in this PR.

Reply:
Thanks to your explanation I'm able to understand the problem.

Now I have a question in your explanation.

> Or it might be possible to fix it without breaking backward compatibility at all. (I haven't checked it yet.)

What do you mean `breaking backward compatibility`?
Both this change and #4074 doesn't seem break backward compatibility.

Reply:
> Now I have a question in your explanation.
> 
> > Or it might be possible to fix it without breaking backward compatibility at all. (I haven't checked it yet.)
> 
> What do you mean `breaking backward compatibility`? Both this change and #4074 doesn't seem break backward compatibility.

This is for the following.

> If so, I thought we could modify the tests and limit the type to `Element`.

If we limit the type of `conf` argument of `configure()`, it will break the backward compatibility.
Some tests, including 3rd-party plugins, may stop working.

So I said

> But we should consider whether it is acceptable to break backward compatibility, even though this is a test-only impact.

Reply:
> If so, I thought we could modify the tests and limit the type to `Element`.

Oh, I understand the point. This sentence was hard to understand. Sorry.
This means the following 2 points.

1. We could modify the tests that passing `Hash` to `Fluent::Plugin::Base::configure()`
1. We could limit tye type of `conf` argument of the `Fluent::Plugin::Base::configure()`

`2.` can break the backward compatibility of some tests including 3rd-party plugins.

Reply:
You mean the code completely rejecting non `Fluent::Config::Element` like the following breaks compatibility?
```ruby
def configure(conf)
  raise SomeError if conf.is_a?(Fluent::Config::Element)
```

Reply:
Yes. I thought we could remove `if conf.is_a?(Fluent::Config::Element)` check in this PR.
This will raise `NoMethodError` when it is passed `Hash`.

```diff
def configure(conf)
-  if conf.is_a?(Fluent::Config::Element)
    if (Fluent::Engine.supervisor_mode && !conf.target_worker_ids.empty?) || conf.for_this_worker?
```

Reply:
Yup, I think breaking compatibility doesn't good idea.
Even if we merge #4074, 3rd party plugins might use obsolete way.

Reply:
Thanks for your advice!
OK! We shouldn't break compatibility even if the impact is only in tests!

> Or it might be possible to fix it without breaking backward compatibility at all.

About this, I think fixing these logic  may allow us to limit the type without breaking compatibility.

https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/compat/parser.rb#L124-L140

However, perhaps we should consider this separately from this PR.

Reply:
> I would like to figure out what we should do about these problems in the future and then figure out how far we should go in this PR.

The reason why I was thinking about limiting the type is that `workers` would not be overwritten unless the type is `Element`.
I think this is a bug of #1507.
When `Hash` is passed to `configure()`, this logic can't overwrite `workers`, and this can cause problems depending on the design of the plugin.

The currently intended specification is a bit unclear, but I think the correct specification is as follows.

* Only the `Fluent::Config::Element` should be passed to `Fluent::Plugin::Base::configure()`.
* For the v0.12 compatible feature, `Hash` can be passed to plugins in the `Fluent::Compat` layer, but it should be converted to the `Element` before being passed to `Fluent::Plugin::Base::configure()`.
* The test codes that pass `Hash` to `configure` of a plugin outside of the `Fluent::Compat` layer are incorrect, and should fail.

Or

* Even if `Hash` is passed to `Fluent::Plugin::Base::configure()`, `workers` is overwritten correctly.

I don't know much about the v0.12 compatible feature, so  I can miss something...
If so, please let me know.

Anyway, I hope we do the following in this PR.

* Clarify the correct specification.
* Leave the appropriate code comments or something for future development.

Reply:
I have looked into the history of past revisions.

It seems that this problem is **NOT** related to the v0.12 compatible feature.

`Fluent::Config::Element` was implemented at the beginning of Fluentd and I can't find any specification that says we can use a `Hash` for `configure()`.

I think the trigger is this fix.

* #103

At this point, there was no `TestDriver` for `Parser`, and it seems that `Hash` was unintentionally passed to `configure()` .

I guess they didn't notice the problem because the other test passed `Hash` to `TestDriver` and it didn't look wrong at first glance.

`ParserTestDriver` is implemented in the following PR, but somehow some tests remained unfixed.

* #702

The same phenomenon is observed for `Formatter`.

* #712

I think these tests misled us and made us think that `Hash` could be passed on.

Reply:
So, I think the correct specification is as follows.

* Only the `Fluent::Config::Element` should be passed to `Fluent::Plugin::Base::configure()`.
* All tests passing `Hash` directly to `Fluent::Plugin::Base::configure()` are wrong and must use `TestDriver`.

Reply:
So I want to limit the type to `Fluent::Config::Element`, and fix some tests in #4074.

On the other hand, even if this was not the intended specification, I think there could be a decision to prioritize maintaining compatibility.
We need to discuss this point carefully.

What do you think? @ashie

Reply:
Thank you for digging the history, it's very helpful to understand the issue.
Now I've found one more interesting thing.
test/test_buffer.rb had also `configure({})` in v0.14.0.pre.1 but it's replaced with `configure(config_element(...))` in test/plugin/test_buffer.rb of v0.14.0.rc1. test/plugin/test_buf_file.rb has also similar history.

Now I think these facts implies that your claim is reasonable.
O.K, we should pass only `Fluent::Config::Element` to `Fluent::Plugin::Base::configure()`.

@abetomo Could you modify your patch according to this policy?

Reply:
Thanks, @ashie !!

Reply:
Thank you for the detailed information.
I will make the changes.

Reply:
Am I correct in my understanding that compatibility will not be maintained?

* Argument of `Fluent::Plugin::Base::configure()` is only `Fluent::Config::Element`.
* Passing `Hash` to `Fluent::Plugin::Base::configure()` makes an error.

Reply:
Yes. Passing Hash is not intented specification. We can assume this argument is always Config::Element.

Reply:
For example, I think simply removing the check of `if conf.is_a?(Fluent::Config::Element)` is fine.

Reply:
I thought it was difficult to understand the cause of the error with `NoMethodError`, so I changed it to throw `ArgumentError`.

Reply:
Nice idea!